enterprise/server/invocation_stat_service/invocation_stat_service.go

```diff
@@ -485,6 +795,211 @@ func (i *InvocationStatService) GetInvocationStat(ctx context.Context, req *inpb
+		return "", nil, err
+	}
+
+	drilldownStr, drilldownArgs, err := getDrilldownQueryFilter(req.GetFilter())
+	if err != nil {
+		return "", nil, err
+	}
+
+	whereString, whereArgs := placeholderQuery.Build()
+
+	args := make([]interface{}, 0)
+	queries := make([]string, 0)
+	subQStr, subQArgs := i.getDrilldownSubquery(ctx, drilldownFields, req, whereString, whereArgs, drilldownStr, drilldownArgs, "")
+	queries = append(queries, subQStr)
+	args = append(args, subQArgs...)
+	for _, s := range drilldownFields {
+		subQStr, subQArgs := i.getDrilldownSubquery(ctx, drilldownFields, req, whereString, whereArgs, drilldownStr, drilldownArgs, s)
+		queries = append(queries, subQStr)
+		args = append(args, subQArgs...)
+	}
+
+	return fmt.Sprintf("SELECT * FROM (%s) ORDER BY totals_first", strings.Join(queries, " UNION ALL ")), args, nil
+}
+
+func addOutputChartEntry(m map[inpb.AggType]*stpb.DrilldownChart, dm map[inpb.AggType]float64, aggType inpb.AggType, label *string, inverse int64, selection int64, totalInBase int64, totalInSelection int64) {
+	chart, exists := m[aggType]
+	if !exists {
+		chart = &stpb.DrilldownChart{}
+		chart.DrilldownDimension = aggType
+		m[aggType] = chart
+		dm[aggType] = -math.MaxFloat64
+	}
+	dm[aggType] = math.Max(dm[aggType], math.Abs(float64(selection)/float64(totalInSelection)-float64(inverse)/float64(totalInBase)))
+	chart.Entry = append(chart.Entry, &stpb.DrilldownEntry{Label: *label, BaseValue: inverse, SelectionValue: selection})
+}
+
+func sortDrilldownChartKeys(dm map[inpb.AggType]float64) *[]inpb.AggType {
+	type pair struct {
+		a inpb.AggType
+		v float64
+	}
+	slice := make([]pair, 0)
+	for k, v := range dm {
+		slice = append(slice, pair{k, v})
+	}
+	sort.SliceStable(slice, func(a, b int) bool {
+		return slice[a].v >= slice[b].v
+	})
+
+	result := make([]inpb.AggType, len(slice))
+	for v, i := range slice {
+		result[v] = i.a
+	}
+	return &result
+}
+
+func (i *InvocationStatService) GetStatDrilldown(ctx context.Context, req *stpb.GetStatDrilldownRequest) (*stpb.GetStatDrilldownResponse, error) {
+	if !config.TrendsHeatmapEnabled() {
+		return nil, status.UnimplementedError("Stat heatmaps are not enabled.")
+	}
+	if !i.isOLAPDBEnabled() {
+		return nil, status.UnimplementedError("Time series charts require using an OLAP DB, but none is configured.")
+	}
+	if err := validateAccessForStats(ctx, i.env, req.GetRequestContext().GetGroupId()); err != nil {
+		return nil, err
+	}
+
+	qStr, qArgs, err := i.getDrilldownQuery(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var rows *sql.Rows
+	rows, err = i.olapdbh.DB(ctx).Raw(qStr, qArgs...).Rows()
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	rsp := &stpb.GetStatDrilldownResponse{}
+
+	rsp.Chart = make([]*stpb.DrilldownChart, 0)
+	m := make(map[inpb.AggType]*stpb.DrilldownChart)
+	dm := make(map[inpb.AggType]float64)
+	type queryOut struct {
+		GormUser       *string
+		GormHost       *string
+		GormRepoUrl    *string
```

Comment:
nit: URL for go conventions. see https://github.com/golang/go/wiki/CodeReviewComments#initialisms